redis_feature_store.js

```diff
@@ -1,60 +1,38 @@
-const redis = require('redis'),
-  dataKind = require('launchdarkly-node-server-sdk/versioned_data_kind'),
-  CachingStoreWrapper = require('launchdarkly-node-server-sdk/caching_store_wrapper');
+const base = require('./redis_base');
+const dataKind = require('launchdarkly-node-server-sdk/versioned_data_kind');
+const CachingStoreWrapper = require('launchdarkly-node-server-sdk/caching_store_wrapper');
 
 const noop = function() {};
 
-function RedisFeatureStore(redisOpts, cacheTTL, prefix, logger, preconfiguredClient) {
+function RedisFeatureStore(redisOpts, cacheTTL, prefix, logger, client) {
+  let options;
+  if (cacheTTL || prefix || logger || client) {
```

Comment:
`logger` is still a separate argument from `redisOpts` from what I can tell, so shouldn't be included in this `if`.

Reply:
_All_ of those arguments are separate arguments. The intention here is to handle two different scenarios:

1. The caller is using the newer semantics, in which you pass just one parameter. That parameter is an object that we're describing in the TypeScript definitions as `LDRedisOptions`; its properties are called `redisOpts`, `cacheTTL`, `prefix`, `logger`, and `client`.

2. Or, the caller is using the older semantics, where `redisOpts`, `cacheTTL`, `prefix`, `logger`, and `client` are separate parameters. The TS type of `redisOpts` in that case is `redis.ClientOptions` (a type from the redis library, not ours).

So, if any parameters besides the first one have values, then we can assume it is (2); that's what the if test is here. And if there's only one parameter, we have to check its properties to see if it is an `LDRedisOptions` or a `redis.ClientOptions`; that's the if test on line 15.

Reply:
I realize that this will be a new major version so we don't absolutely need to go to the effort of being backward-compatible like this— I could have just changed it to support the new parameter only. But I'm trying to minimize the impact on people who aren't using the big segments feature, until the next major version of the SDK where there will be bigger configuration changes.

Reply:
I don't think the code is consistent currently whether the intent was `logger` is part of `redisOpts` or a separate argument. The implementation in this PR seems like it would be consistent with `logger` not being part of `redisOpts` when calling with the new semantics, if this `if` didn't check for it. But now that I'm thinking more about it, it makes more sense as part of `LDRedisOptions` for when we remove this compatibility at some point in the future. I'll leave review comments for where I think `logger` as part of `LDRedisOptions` doesn't seem right in this current implementation.

Reply:
I think I am still failing to explain this, and I also did make one error above. Either that, or I'm failing to understand what you mean by "logger being part of redisOpts".

In scenario 1 above, `redisOpts` is a `redis.ClientOptions` object. It contains only Redis client options, not any of our own SDK-related properties.

In scenario 2, `redisOpts` is an `LDRedisOptions` object. It _contains_ the `redis.ClientOptions` object in a `redisOpts` property of its own. Contrary to what I said above, it does not have a `logger` property _because that is going to be set from the SDK's config.logger_.

The only reason you would ever pass a separate `logger` parameter is if you are using the old calling semantics.

Reply:
I've looked over this code 5 times now and it is definitely doing what it was intended to do. This is just a problem of communication and/or documentation. If there's a way to improve the comments to better convey what I'm saying here, please let me know.